Refactor project structure and add Binance data dumper

river-walras opened the pull request:
- Introduced a new `main.py` file to serve as the entry point for the application.
- Updated `pyproject.toml` to reflect new project name and versioning.
- Removed `requirements.txt` in favor of using Poetry for dependency management.
- Enhanced `.gitignore` to exclude cache files and virtual environments.
- Cleaned up `poetry.lock` to align with the new dependencies.
- using `parquet` instead of `csv`
```
            df = pd.read_csv(path_zip_raw_file)
            path_parquet_file = path_zip_raw_file.replace(".zip", ".parquet")
            df.to_parquet(path_parquet_file)
```

river-walras committed:
Refactor project structure and add Binance data dumper

- Introduced a new `main.py` file to serve as the entry point for the application.
- Added `data_dumper.py` to implement the `BinanceDataDumper` class for downloading historical data from Binance.
- Created `logger.py` for logging setup and management.
- Updated `pyproject.toml` to reflect new project name and versioning.
- Removed `requirements.txt` in favor of using Poetry for dependency management.
- Enhanced `.gitignore` to exclude cache files and virtual environments.
- Cleaned up `poetry.lock` to align with the new dependencies.

river-walras committed:
Update data handling and dependencies

- Modified `main.py` to dynamically fetch trading pairs for data dumping instead of hardcoding 'ETHUSDT'.
- Updated `pyproject.toml` to include `certifi` as a dependency for SSL certificate management.
- Enhanced `data_dumper.py` to utilize `certifi` for creating a secure SSL context when fetching data from Binance.
- Updated `poetry.lock` to reflect the new dependency and ensure compatibility.

river-walras committed:
Remove main.py as it is no longer needed for the application entry point. This file contained the logic for fetching and dumping Binance trading data, which has been refactored or replaced in the project structure.

river-walras committed:
Update .gitignore and enhance data_dumper.py for improved data handling

- Changed the data export format in data_dumper.py from CSV to Parquet for better performance and storage efficiency.
- Implemented a fix for handling spot klines data formatting issues after 2025-01-01, ensuring correct column names during CSV reading.

stas-prokopiev reviewed (commented):
Hello, 
Thank you for your interest in the project,
I like what you are doing and parquet is obviously significantly more efficient

Could you please add additional argument for user to choose data type of saved/loaded data to work with (with default value "csv") ?
It's needed for backward compatibility.
And change the package version from 0.1.0 -> 0.2.0
After it I'll be able to merge your changes and upload the updated version on PYPI

Best Regards,
Stas.

river-walras commented:
> Hello, Thank you for your interest in the project, I like what you are doing and parquet is obviously significantly more efficient
> 
> Could you please add additional argument for user to choose data type of saved/loaded data to work with (with default value "csv") ? It's needed for backward compatibility. And change the package version from 0.1.0 -> 0.2.0 After it I'll be able to merge your changes and upload the updated version on PYPI
> 
> Best Regards, Stas.

Done. Added the additional argument and change the version to 0.2.0.

river-walras committed:
Enhance BinanceDataDumper to support multiple save formats

- Added support for saving data in both CSV and Parquet formats, allowing users to choose their preferred format.
- Updated the constructor to include a new parameter `save_format` with validation against allowed formats.
- Modified data extraction logic to handle the chosen save format during data dumping.
- Ensured backward compatibility by defaulting to CSV format if no format is specified.

river-walras committed:
Update version to 0.2.0 in pyproject.toml for release

Commit:
Enhance download functionality in BinanceDataDumper with retry mechanism

- Added retry logic to the `_download_raw_file` method to handle transient download errors, allowing up to 5 retries with exponential backoff.
- Introduced parameters `retry_base_time` and `max_retries` to customize retry behavior.
- Improved error handling for various HTTP and URL errors, providing detailed logging for failures.
- Ensured that the download process remains robust against temporary network issues.

river-walras commented:
Hi,

I have added a retry logic for `_download_raw_file`. 

- Added retry logic to the `_download_raw_file` method to handle transient download errors, allowing up to 5 retries with exponential backoff.
- Improved error handling for various HTTP and URL errors, providing detailed logging for failures.
- Ensured that the download process remains robust against temporary network issues.
- updated `_get_list_all_available_files` to `_get_list_all_available_files_v2` using `boto3` 
- added `_download_raw_file_v2` downloaded data directly from s3 bucket which is more stable

Best,
River

river-walras committed:
Update dependencies and enhance BinanceDataDumper for S3 integration

- Added `boto3` and `botocore` as dependencies in `pyproject.toml` to facilitate AWS S3 interactions.
- Enhanced `BinanceDataDumper` to include S3 client configuration for downloading data from an S3 bucket.
- Implemented a new method `_get_list_all_available_files_v2` to retrieve available files from S3, improving data fetching capabilities.
- Updated `poetry.lock` to reflect the new dependencies and ensure compatibility with the latest package versions.

Commit:
Enhance BinanceDataDumper with improved S3 file download functionality

- Added a new method `_download_raw_file_v2` to download files from S3 with enhanced error handling and retry logic.
- Integrated progress tracking for downloading 'trades' data using a progress bar.
- Improved logging for file not found and download errors, ensuring better visibility into download issues.
- Utilized `botocore` for handling S3 client errors effectively.

Commit:
Refactor BinanceDataDumper to enhance download reliability and configuration

- Introduced retry logic in both `_download_raw_file` and `_download_raw_file_v2` methods, utilizing instance variables for `retry_base_time` and `max_retries`.
- Updated the constructor to accept `retry_base_time` and `max_retries` parameters, improving customization for download attempts.
- Ensured consistent retry behavior across different download methods, enhancing robustness against transient errors.

stas-prokopiev commented:
Hi,
It's a pity that you decided to close this pull request,
You've added a lot of valuable features.
Can I at lease merge what you've done until this moment?
Can I mention you at README as a contributor?

river-walras commented:
> Hi, It's a pity that you decided to close this pull request, You've added a lot of valuable features. Can I at lease merge what you've done until this moment? Can I mention you at README as a contributor?
Hi, thanks for mentioning me as a contributor.

I've identified a significant issue in the code, so I've decided to close this pull request for now.

Have you encountered this situation which might cause some error?

Lets say Today's date is January 22, 2025, but the latest available month is only November 2024, even though December 2024 is complete.

```python
        for ticker in tqdm(list_trading_pairs, leave=True, desc="Tickers"):
            # 1) Download all monthly data
            if self._data_type != "metrics" and (
                date_end_first_day_of_month - relativedelta(days=1) > date_start
            ):
                self._download_data_for_1_ticker(
                    ticker=ticker,
                    date_start=date_start,
                    date_end=(date_end_first_day_of_month - relativedelta(days=1)), # 2024-12-31, However, the data in December is missing, so it only download until 2024-11-30, the whole data in December would missing
                    timeperiod_per_file="monthly",
                    is_to_update_existing=is_to_update_existing,
                )
            # 2) Download all daily date
            if self._data_type == "metrics":
                date_start_daily = date_start
            else:
                date_start_daily = date_end_first_day_of_month
            self._download_data_for_1_ticker(
                ticker=ticker,
                date_start=date_start_daily, # The date start should begin at 2024-12-01 instead of 2025-01-01
                date_end=date_end,
                timeperiod_per_file="daily",
                is_to_update_existing=is_to_update_existing,
            )
 ```
